File: src/ui/widgets/rounded_overlay.py
from __future__ import annotations

from PySide6.QtCore import Qt
from PySide6.QtGui import QColor, QPainter
from PySide6.QtWidgets import QWidget


class RoundedOverlayWidget(QWidget):
    """Child widget that paints its own AA rounded background.

    Works correctly when hosted over a QOpenGLWidget where QSS border-radius
    on a child widget leaves a leaking rectangle outside the rounded shape.
    """

    def __init__(self, parent=None, *, bg_color=QColor(0, 0, 0, 140), radius=6):
        super().__init__(parent)
        self.setAttribute(Qt.WidgetAttribute.WA_NoSystemBackground, True)
        self.setAttribute(Qt.WidgetAttribute.WA_TranslucentBackground, True)
        self._bg_color = bg_color
        self._radius = float(radius)

    def paintEvent(self, event):
        painter = QPainter(self)
        painter.setRenderHint(QPainter.RenderHint.Antialiasing, True)
        painter.setCompositionMode(QPainter.CompositionMode.CompositionMode_Source)
        painter.fillRect(self.rect(), QColor(0, 0, 0, 0))
        painter.setCompositionMode(QPainter.CompositionMode.CompositionMode_SourceOver)
        painter.setPen(Qt.PenStyle.NoPen)
        painter.setBrush(self._bg_color)
        painter.drawRoundedRect(self.rect(), self._radius, self._radius)
        painter.end()

File: cpp/app/ui/widgets/rounded_overlay.cpp
#include "ui/widgets/rounded_overlay.h"

#include <QPainter>

namespace imgsli::app::ui::widgets {

RoundedOverlayWidget::RoundedOverlayWidget(QWidget* parent, QColor bgColor,
                                           qreal radius)
    : QWidget(parent), bgColor_(std::move(bgColor)), radius_(radius) {
  setAttribute(Qt::WA_NoSystemBackground, true);
  setAttribute(Qt::WA_TranslucentBackground, true);
}

void RoundedOverlayWidget::setBackgroundColor(const QColor& color) {
  bgColor_ = color;
  update();
}

void RoundedOverlayWidget::setRadius(qreal radius) {
  radius_ = radius;
  update();
}

void RoundedOverlayWidget::paintEvent(QPaintEvent* /*event*/) {
  QPainter painter(this);
  painter.setRenderHint(QPainter::Antialiasing, true);
  painter.setCompositionMode(QPainter::CompositionMode_Source);
  painter.fillRect(rect(), QColor(0, 0, 0, 0));
  painter.setCompositionMode(QPainter::CompositionMode_SourceOver);
  painter.setPen(Qt::NoPen);
  painter.setBrush(bgColor_);
  painter.drawRoundedRect(rect(), radius_, radius_);
}

}  // namespace imgsli::app::ui::widgets
